File: src/main/java/io/github/faecraft/gentlefawn/block/model/BuckHeadModel.java
package io.github.faecraft.gentlefawn.block.model;

import net.minecraft.client.model.ModelPart;
import net.minecraft.client.render.VertexConsumer;
import net.minecraft.client.render.entity.model.SkullEntityModel;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.util.math.Quaternion;

public class BuckHeadModel extends SkullEntityModel {
    private final ModelPart head;
    private final ModelPart horns2;
    private final ModelPart bone9;
    private final ModelPart bone10;
    private final ModelPart bone11;
    private final ModelPart bone12;
    private final ModelPart bone13;
    private final ModelPart bone14;
    private final ModelPart bone8;
    private final ModelPart bone7;

    public BuckHeadModel() {
        textureWidth = 72;
        textureHeight = 84;
        head = new ModelPart(this);
        head.setPivot(0.0F, 16.0F, 0.0F);
        head.setTextureOffset(24, 36).addCuboid(-6.0F, -4.0F, -20.0F, 12.0F, 12.0F, 12.0F, 0.0F, true);
        head.setTextureOffset(0, 0).addCuboid(-9.0F, -10.0F, -8.0F, 18.0F, 18.0F, 18.0F, 0.0F, true);

        horns2 = new ModelPart(this);
        horns2.setPivot(0.0F, -17.0F, 6.0F);
        head.addChild(horns2);


        bone9 = new ModelPart(this);
        bone9.setPivot(-2.0F, -2.75F, -0.5F);
        horns2.addChild(bone9);
        bone9.setTextureOffset(45, 63).addCuboid(-3.0F, 3.75F, 0.0F, 3.0F, 6.0F, 0.0F, 0.0F, false);
        bone9.setTextureOffset(27, 66).addCuboid(-3.0F, -5.25F, 0.0F, 3.0F, 3.0F, 0.0F, 0.0F, false);

        bone10 = new ModelPart(this);
        bone10.setPivot(2.0F, -2.75F, -0.5F);
        horns2.addChild(bone10);
        bone10.setTextureOffset(27, 66).addCuboid(-0.1F, -5.25F, 0.0F, 3.0F, 3.0F, 0.0F, 0.0F, false);
        bone10.setTextureOffset(45, 63).addCuboid(0.0F, 3.75F, 0.0F, 3.0F, 6.0F, 0.0F, 0.0F, false);

        bone11 = new ModelPart(this);
        bone11.setPivot(-4.1143F, -5.7857F, -0.9F);
        horns2.addChild(bone11);
        setRotationAngle(bone11, 0.0F, -0.2618F, 0.0F);
        bone11.setTextureOffset(27, 60).addCuboid(-6.7357F, -20.2143F, 0.6F, 3.0F, 3.0F, 0.0F, 0.0F, false);
        bone11.setTextureOffset(36, 66).addCuboid(-9.7357F, -17.2143F, 0.6F, 3.0F, 6.0F, 0.0F, 0.0F, false);
        bone11.setTextureOffset(36, 66).addCuboid(-6.7357F, -8.2143F, 0.6F, 3.0F, 6.0F, 0.0F, 0.0F, false);
        bone11.setTextureOffset(27, 72).addCuboid(-9.7357F, -5.2143F, 0.6F, 3.0F, 3.0F, 0.0F, 0.0F, false);
        bone11.setTextureOffset(36, 60).addCuboid(-6.7357F, -2.2143F, 0.6F, 3.0F, 3.0F, 0.0F, 0.0F, false);
        bone11.setTextureOffset(45, 60).addCuboid(-3.7357F, 0.7857F, 0.6F, 3.0F, 9.0F, 0.0F, 0.0F, false);
        bone11.setTextureOffset(36, 75).addCuboid(-9.7357F, 3.7857F, 0.6F, 6.0F, 3.0F, 0.0F, 0.0F, false);

        bone12 = new ModelPart(this);
        bone12.setPivot(4.1143F, -5.7857F, -0.8F);
        horns2.addChild(bone12);
        setRotationAngle(bone12, 0.0F, 0.2618F, 0.0F);
        bone12.setTextureOffset(27, 60).addCuboid(3.7357F, -20.2143F, 0.5F, 3.0F, 3.0F, 0.0F, 0.0F, false);
        bone12.setTextureOffset(36, 66).addCuboid(6.7357F, -17.2143F, 0.5F, 3.0F, 6.0F, 0.0F, 0.0F, false);
        bone12.setTextureOffset(27, 72).addCuboid(6.7357F, -5.2143F, 0.5F, 3.0F, 3.0F, 0.0F, 0.0F, false);
        bone12.setTextureOffset(36, 66).addCuboid(3.7357F, -8.2143F, 0.5F, 3.0F, 6.0F, 0.0F, 0.0F, false);
        bone12.setTextureOffset(36, 60).addCuboid(3.7357F, -2.2143F, 0.5F, 3.0F, 3.0F, 0.0F, 0.0F, false);
        bone12.setTextureOffset(45, 60).addCuboid(0.7357F, 0.7857F, 0.5F, 3.0F, 9.0F, 0.0F, 0.0F, false);
        bone12.setTextureOffset(36, 81).addCuboid(3.7357F, 3.7857F, 0.5F, 6.0F, 3.0F, 0.0F, 0.0F, false);

        bone13 = new ModelPart(this);
        bone13.setPivot(-6.575F, -6.375F, -1.8F);
        horns2.addChild(bone13);
        setRotationAngle(bone13, 0.0F, -0.4363F, 0.0F);
        bone13.setTextureOffset(36, 75).addCuboid(-12.825F, -1.625F, 2.05F, 6.0F, 3.0F, 0.0F, 0.0F, false);
        bone13.setTextureOffset(27, 60).addCuboid(-12.825F, -13.625F, 2.05F, 3.0F, 3.0F, 0.0F, 0.0F, false);
        bone13.setTextureOffset(27, 66).addCuboid(-15.825F, -4.625F, 2.05F, 3.0F, 3.0F, 0.0F, 0.0F, false);
        bone13.setTextureOffset(36, 66).addCuboid(-9.825F, -10.625F, 2.05F, 3.0F, 6.0F, 0.0F, 0.0F, false);

        bone14 = new ModelPart(this);
        bone14.setPivot(6.4F, -6.7F, -1.8F);
        horns2.addChild(bone14);
        setRotationAngle(bone14, 0.0F, 0.5236F, 0.0F);
        bone14.setTextureOffset(36, 81).addCuboid(6.8F, -1.3F, 2.7F, 6.0F, 3.0F, 0.0F, 0.0F, false);
        bone14.setTextureOffset(36, 66).addCuboid(6.8F, -10.3F, 2.7F, 3.0F, 6.0F, 0.0F, 0.0F, false);
        bone14.setTextureOffset(27, 60).addCuboid(9.8F, -13.3F, 2.7F, 3.0F, 3.0F, 0.0F, 0.0F, false);
        bone14.setTextureOffset(27, 66).addCuboid(12.8F, -4.3F, 2.7F, 3.0F, 3.0F, 0.0F, 0.0F, false);

        bone8 = new ModelPart(this);
        bone8.setPivot(-11.5F, -10.0F, 5.5F);
        head.addChild(bone8);
        setRotationAngle(bone8, 0.0F, 0.0F, -1.0472F);
        bone8.setTextureOffset(0, 57).addCuboid(-4.5F, -9.0F, -1.5F, 9.0F, 18.0F, 3.0F, 0.0F, true);

        bone7 = new ModelPart(this);
        bone7.setPivot(11.5F, -10.0F, 5.5F);
        head.addChild(bone7);
        setRotationAngle(bone7, 0.0F, 0.0F, 1.0472F);
        bone7.setTextureOffset(0, 36).addCuboid(-4.5F, -9.0F, -1.5F, 9.0F, 18.0F, 3.0F, 0.0F, true);
    }

    @Override
    public void render(MatrixStack matrices, VertexConsumer vertices, int light, int overlay, float red, float green, float blue, float alpha) {
        matrices.push();

        matrices.translate(0.0, -0.75, 0.0);
        matrices.scale(0.5F, 0.5F, 0.5F);
        matrices.multiply(new Quaternion(0.0F, 180.0F, 0.0F, true));

        head.render(matrices, vertices, light, overlay, red, green, blue, alpha);

        matrices.pop();
    }

    @Override
    public void method_2821(float f, float g, float h) {
        head.yaw = (g + 180) * (float) (Math.PI / 180F);
        head.pitch = h * (float) (Math.PI / 180F);
    }

    public void setRotationAngle(ModelPart bone, float x, float y, float z) {
        bone.pitch = x;
        bone.yaw = y;
        bone.roll = z;
    }
}

File: src/main/kotlin/io/github/faecraft/gentlefawn/block/model/BuckHeadModel.kt
package io.github.faecraft.gentlefawn.block.model

import net.minecraft.client.model.ModelPart
import net.minecraft.client.render.VertexConsumer
import net.minecraft.client.render.entity.model.SkullEntityModel
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.util.math.Quaternion

class BuckHeadModel() : SkullEntityModel() {
    private val head = ModelPart(this)
    private val horns = ModelPart(this)

    private val ear1 = ModelPart(this)
    private val ear2 = ModelPart(this)

    private val part1 = ModelPart(this)
    private val part2 = ModelPart(this)
    private val part3 = ModelPart(this)
    private val part4 = ModelPart(this)
    private val part5 = ModelPart(this)
    private val part6 = ModelPart(this)

    init {
        textureWidth = 72
        textureHeight = 84

        head.addChild(ear1)
        head.addChild(ear2)
        head.addChild(horns)

        horns.addChild(part1)
        horns.addChild(part2)
        horns.addChild(part3)
        horns.addChild(part4)
        horns.addChild(part5)
        horns.addChild(part6)

        setAngles(ear1, 0.0F, 0.0F, -1.0472F)
        setAngles(ear2, 0.0F, 0.0F, 1.0472F)

        setAngles(part3, 0.0F, -0.2618F, 0.0F)
        setAngles(part4, 0.0F, 0.2618F, 0.0F)
        setAngles(part5, 0.0F, -0.4363F, 0.0F)
        setAngles(part6, 0.0F, 0.5236F, 0.0F)

        head.setPivot(0.0F, 16.0F, 0.0F)
        head.setTextureOffset(24, 36)
            .addCuboid(-6.0F, -4.0F, -20.0F, 12.0F, 12.0F, 12.0F, 0.0F, true)
        head.setTextureOffset(0, 0)
            .addCuboid(-9.0F, -10.0F, -8.0F, 18.0F, 18.0F, 18.0F, 0.0F, true)

        horns.setPivot(0.0F, -17.0F, 6.0F)

        part1.setPivot(-2.0F, -2.75F, -0.5F)
        part1.setTextureOffset(45, 63)
            .addCuboid(-3.0f, 3.75f, 0.0f, 3.0f, 6.0f, 0.0f, 0.0f, false)
        part1.setTextureOffset(27, 66)
            .addCuboid(-3.0f, -5.25f, 0.0f, 3.0f, 3.0f, 0.0f, 0.0f, false)
        
        part2.setPivot(2.0F, -2.75F, -0.5F)
        part2.setTextureOffset(27, 66)
            .addCuboid(-0.1f, -5.25f, 0.0f, 3.0f, 3.0f, 0.0f, 0.0f, false)
        part2.setTextureOffset(45, 63)
            .addCuboid(0.0f, 3.75f, 0.0f, 3.0f, 6.0f, 0.0f, 0.0f, false)
        
        part3.setPivot(-4.1143f, -5.7857f, -0.9f)
        part3.setTextureOffset(27, 60)
            .addCuboid(-6.7357f, -20.2143f, 0.6f, 3.0f, 3.0f, 0.0f, 0.0f, false)
        part3.setTextureOffset(36, 66)
            .addCuboid(-9.7357f, -17.2143f, 0.6f, 3.0f, 6.0f, 0.0f, 0.0f, false)
        part3.setTextureOffset(36, 66)
            .addCuboid(-6.7357f, -8.2143f, 0.6f, 3.0f, 6.0f, 0.0f, 0.0f, false)
        part3.setTextureOffset(27, 72)
            .addCuboid(-9.7357f, -5.2143f, 0.6f, 3.0f, 3.0f, 0.0f, 0.0f, false)
        part3.setTextureOffset(36, 60)
            .addCuboid(-6.7357f, -2.2143f, 0.6f, 3.0f, 3.0f, 0.0f, 0.0f, false)
        part3.setTextureOffset(45, 60)
            .addCuboid(-3.7357f, 0.7857f, 0.6f, 3.0f, 9.0f, 0.0f, 0.0f, false)
        part3.setTextureOffset(36, 75)
            .addCuboid(-9.7357f, 3.7857f, 0.6f, 6.0f, 3.0f, 0.0f, 0.0f, false)

        part4.setPivot(4.1143F, -5.7857F, -0.8F);
        part4.setTextureOffset(27, 60)
            .addCuboid(3.7357f, -20.2143f, 0.5f, 3.0f, 3.0f, 0.0f, 0.0f, false)
        part4.setTextureOffset(36, 66)
            .addCuboid(6.7357f, -17.2143f, 0.5f, 3.0f, 6.0f, 0.0f, 0.0f, false)
        part4.setTextureOffset(27, 72)
            .addCuboid(6.7357f, -5.2143f, 0.5f, 3.0f, 3.0f, 0.0f, 0.0f, false)
        part4.setTextureOffset(36, 66)
            .addCuboid(3.7357f, -8.2143f, 0.5f, 3.0f, 6.0f, 0.0f, 0.0f, false)
        part4.setTextureOffset(36, 60)
            .addCuboid(3.7357f, -2.2143f, 0.5f, 3.0f, 3.0f, 0.0f, 0.0f, false)
        part4.setTextureOffset(45, 60)
            .addCuboid(0.7357f, 0.7857f, 0.5f, 3.0f, 9.0f, 0.0f, 0.0f, false)
        part4.setTextureOffset(36, 81)
            .addCuboid(3.7357f, 3.7857f, 0.5f, 6.0f, 3.0f, 0.0f, 0.0f, false)

        part5.setPivot(-6.575f, -6.375f, -1.8f)
        part5.setTextureOffset(36, 75)
            .addCuboid(-12.825f, -1.625f, 2.05f, 6.0f, 3.0f, 0.0f, 0.0f, false)
        part5.setTextureOffset(27, 60)
            .addCuboid(-12.825f, -13.625f, 2.05f, 3.0f, 3.0f, 0.0f, 0.0f, false)
        part5.setTextureOffset(27, 66)
            .addCuboid(-15.825f, -4.625f, 2.05f, 3.0f, 3.0f, 0.0f, 0.0f, false)
        part5.setTextureOffset(36, 66)
            .addCuboid(-9.825f, -10.625f, 2.05f, 3.0f, 6.0f, 0.0f, 0.0f, false)

        part6.setPivot(6.4f, -6.7f, -1.8f)
        part6.setTextureOffset(36, 81)
            .addCuboid(6.8f, -1.3f, 2.7f, 6.0f, 3.0f, 0.0f, 0.0f, false)
        part6.setTextureOffset(36, 66)
            .addCuboid(6.8f, -10.3f, 2.7f, 3.0f, 6.0f, 0.0f, 0.0f, false)
        part6.setTextureOffset(27, 60)
            .addCuboid(9.8f, -13.3f, 2.7f, 3.0f, 3.0f, 0.0f, 0.0f, false)
        part6.setTextureOffset(27, 66)
            .addCuboid(12.8f, -4.3f, 2.7f, 3.0f, 3.0f, 0.0f, 0.0f, false)


        ear1.setPivot(-11.5F, -10.0F, 5.5F)
        ear1.setTextureOffset(0, 57)
            .addCuboid(-4.5F, -9.0F, -1.5F, 9.0F, 18.0F, 3.0F, 0.0F, true)

        ear2.setPivot(11.5F, -10.0F, 5.5F)
        ear2.setTextureOffset(0, 36)
            .addCuboid(-4.5F, -9.0F, -1.5F, 9.0F, 18.0F, 3.0F, 0.0F, true)
    }

    override fun render(
        matrices: MatrixStack,
        vertices: VertexConsumer,
        light: Int,
        overlay: Int,
        red: Float,
        green: Float,
        blue: Float,
        alpha: Float
    ) {
        matrices.push()

        matrices.translate(0.0, -0.75, 0.0)
        matrices.scale(0.5F, 0.5F, 0.5F)
        matrices.multiply(Quaternion(0.0F, 180.0F, 0.0F, true))

        head.render(matrices, vertices, light, overlay, red, green, blue, alpha)

        matrices.pop()
    }

    override fun method_2821(f: Float, g: Float, h: Float) {
        head.yaw = (g + 180) * (Math.PI / 180F).toFloat()
        head.pitch = h * (Math.PI / 180F).toFloat()
    }

    private fun setAngles(part: ModelPart, pitch: Float, yaw: Float, roll: Float) {
        // NOTE: This auto-generated method will likely never be changed, so don't bother copying it in
        // from generated model code.

        part.pitch = pitch
        part.yaw = yaw
        part.roll = roll
    }
}
